Sources/HummingbirdCore/Request/RequestBody+inboundClose.swift

```diff
@@ -0,0 +1,153 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import HTTPTypes
+import NIOConcurrencyHelpers
+import NIOCore
+import NIOHTTPTypes
+
+#if compiler(>=6.0)
+@available(macOS 15, iOS 18, tvOS 18, *)
+extension RequestBody {
+    /// Run provided closure but cancel it if the inbound request part stream is closed.
+    ///
+    /// This function is designed for use with long running requests like server sent events. It assumes you
+    /// are not going to be using the request body after calling as it consumes the request body, it also assumes
+    /// you havent edited the request body prior to calling this function.
+    ///
+    /// If the response finishes the connection will be closed.
+    ///
+    /// - Parameters
+    ///   - operation: The actual operation
+    ///   = onInboundClose: handler invoked when inbound is closed
+    /// - Returns: Return value of operation
+    @available(macOS 15, iOS 18, tvOS 18, *)
+    public func consumeWithInboundCloseHandler<Value: Sendable>(
+        _ operation: sending (RequestBody) async throws -> Value,
+        onInboundClosed: @Sendable @escaping () -> Void
+    ) async throws -> Value {
+        let iterator: UnsafeTransfer<NIOAsyncChannelInboundStream<HTTPRequestPart>.AsyncIterator> =
+            switch self._backing {
+            case .nioAsyncChannelRequestBody(let iterator):
+                iterator.underlyingIterator
+            default:
+                preconditionFailure("Cannot run consumeWithInboundCloseHandler on edited request body")
```

Comment:
I agree in-part, but crashing is a bit of an extreme measure here in my opinion. I'd be more open to an `assertionFailure`, but a precondition is harsh for this type of error.

Reply:
If we do want to keep the preconditionFailure, we really have to add a way for a user to know of a `RequestBody` is still a stream at all. I'd like not to expose that, but some middleware might collate your request body and you don't want to crash because of that.

Reply:
The other option is a keep a reference to the underlying iterator in the other cases so it can still be used

Reply:
That's a good option

Reply:
That doesn't work very well in the end. It's a bit clunky to have to keep a reference around and then you are iterating the wrong sequence for the request body ie the original one and not the one attached to the request now. I can probably add it for requests where the body has been collected into one buffer (as iteration is finished then), but if someone completely replaces the request body stream it just won't work.

Reply:
I have other options for keeping the original async iterator around. But they are quite involved. So will look to do in a separate PR